# Applying Decision Trees

## Introduction to the Data Set
Over the past two missions, we learned about how decision trees are constructed. We used a modified version of ID3, which is a bit simpler than the most common tree building algorithms, C4.5 and CART. The basics are the same, however, so we can apply what we learned about how decision trees work to any tree construction algorithm.<br>

In this mission, we'll learn about when to use decision trees, and how to use them most effectively.<br>

We'll continue using the 1994 census data on U.S. incomes we worked with in the previous mission. It contains information on marital status, age, type of work, and more. The target column, `high_income`, indicates an income of less than or equal to 50k a year (`0`), or more than 50k a year (`1`).<br>

You can download the data from the [University of California, Irvine's website](http://archive.ics.uci.edu/ml/datasets/Adult).

## Using Decision Trees With scikit-learn

We can use the [scikit-learn](http://scikit-learn.org/) package to fit a decision tree. The interface is very similar to other algorithms we've fit in the past.<br>

We use the [DecisionTreeClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn.tree.DecisionTreeClassifier) class for classification problems, and [DecisionTreeRegressor](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html#sklearn.tree.DecisionTreeRegressor) for regression problems. The `sklearn.tree` package includes both of these classes.<br>

In this case, we're predicting a binary outcome, so we'll use a classifier.<br>

The first step is to train the classifier on the data. We'll use the `fit` method on a classifier to do this.

Fit `clf` to the `income` data.

* Pass in `income[columns]` so that we only use the named columns as predictors.
* The target is the `high_income` column.

In [1]:
import pandas as pd
import numpy as np
import math
from sklearn.tree import DecisionTreeClassifier

In [2]:
income = pd.read_csv('data/income.csv')

# A list of columns to train with
columns = ["age", "workclass", "education_num", "marital_status", "occupation", "relationship", "race", "sex", "hours_per_week", "native_country"]
for col in columns+['high_income']:
    converted = pd.Categorical.from_array(income[col])
    income[col] = converted.codes

/usr/local/lib/python3.6/site-packages/ipykernel_launcher.py:6: FutureWarning: Categorical.from_array is deprecated, use Categorical instead
  


In [3]:
income.head()

age  workclass  fnlwgt   education  education_num  marital_status  \
0   22          7   77516   Bachelors             12               4   
1   33          6   83311   Bachelors             12               2   
2   21          4  215646     HS-grad              8               0   
3   36          4  234721        11th              6               2   
4   11          4  338409   Bachelors             12               2   

   occupation  relationship  race  sex  capital_gain  capital_loss  \
0           1             1     4    1          2174             0   
1           4             0     4    1             0             0   
2           6             1     4    1             0             0   
3           6             0     2    1             0             0   
4          10             5     2    0             0             0   

   hours_per_week  native_country  high_income  
0              39              39            0  
1              12              39            0  
2              39              39            0  
3              39              39            0  
4              39               5            0

In [4]:
# Instantiate the classifier
# Set random_state to 1 to make sure the results are consistent
clf = DecisionTreeClassifier(random_state=1)

# We've already loaded the variable "income," which contains all of the income data
clf.fit(income[columns], income['high_income'])

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=1,
            splitter='best')

## Splitting the Data into Train and Test Sets

Now that we've fit a model, we can make predictions. We'll want to split our data into training and testing sets first. If we don't, we'll be making predictions on the same data that we train our algorithm with. This leads to overfitting, and will make our error appear lower than it is.<br>

While we covered overfitting in more depth in previous missions, here's a quick recap. If you memorize how to perform three specific addition problems (`2+2`, `3+6`, `3+3`), you'll get those specific problems correct every time.<br>

On the other hand, if someone asks you what `4+4` is, you won't know how to do it, because you don't know the rules of addition. If you learn the rules of addition, you'll get problems wrong sometimes (because operations like `3443343434+24344343` can be hard to do mentally). Even so, you'll be able to do any problem, and you'll get most of them right.<br>

The first example represents overfitting, where you memorize the details of the training set, but can't generalize to new examples you're asked to make predictions on.<br>

We can avoid overfitting by always making predictions and evaluating error on data that we haven't trained our algorithm with. This will show us when we're overfitting by giving us a realistic error on data that the algorithm hasn't seen before.<br>

We can split the data by shuffling the order of the dataframe, then selecting certain rows to include in the training set, and certain rows to include in the testing set.<br>

In this case, we'll make `80%` of our rows training data, and the rest testing data.

All of the rows in `income` with a position up to `train_max_row` (but not including it) will be part of the training set.

* Make a new dataframe called `train` containing all of these rows.
* Make a dataframe called `test` containing all of the rows with a position greater than or equal to `train_max_row`.

In [5]:
# Set a random seed so the shuffle is the same every time
np.random.seed(1)

# Shuffle the rows  
# This permutes the index randomly using np.random.permutation
# Then, it reindexes the dataframe with the result
# The net effect is to put the rows into random order
income = income.reindex(np.random.permutation(income.index))
train_max_row = math.floor(income.shape[0] * .8)

train = income.iloc[:train_max_row]
test = income.iloc[train_max_row:]

## Evaluating Error with AUC

While there are many methods for evaluating error with classification, we'll use AUC, which we've already covered extensively in the machine learning material. [AUC](https://en.wikipedia.org/wiki/Receiver_operating_characteristic#Area_under_curve) ranges from `0` to `1`, so it's ideal for binary classification. The higher the AUC, the more accurate our predictions.<br>

We can compute AUC with the [roc_auc_score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score) function from `sklearn.metrics`. This function takes in two parameters:
* `y_true`: true labels
* `y_score`: predicted labels

It then calculates and returns the AUC value.

* Compute the AUC between `predictions` and the `high_income` column of `test`, and assign the result to `error`.
* Use the `print` function to display `error`.

In [6]:
from sklearn.metrics import roc_auc_score

clf = DecisionTreeClassifier(random_state=1)
clf.fit(train[columns], train["high_income"])

predictions = clf.predict(test[columns])

In [7]:
error = roc_auc_score(test['high_income'], predictions)
print(error)

0.693568175543


## Computing Error on the Training Set

The AUC for the predictions on the testing set is about `.694`. Let's compare this against the AUC for predictions on the training set to see if the model is overfitting.<br>

**It's normal for the model to predict the training set better than the testing set**. After all, it has full knowledge of that data and the outcomes. 
* However, if the AUC between training set predictions and actual values is significantly higher than the AUC between test set predictions and actual values, it's a sign that the model may be overfitting.

* Print out the AUC score between `predictions` and the `high_income` column of `train`.

In [8]:
predictions = clf.predict(train[columns])
print(roc_auc_score(train['high_income'], predictions))

0.947124450144


## Decision Tree Overfitting

Our AUC on the training set was `.947`, and the AUC on the test set was `.694`. There's no hard and fast rule on when overfitting is occurring, but our model is predicting the training set much better than the test set. Splitting the data into training and testing sets doesn't prevent overfitting -- it just helps us detect and fix it.<br>

Based on our AUC measurements, it appears that we are in fact overfitting. Let's take a closer look at why decision trees might overfit.<br>

In the last mission, we looked at this data:

```python
high_income    age    marital_status
0              20     0
0              60     2
0              40     1
1              25     1
1              35     2
1              55     1
```

Here's the full diagram for the decision tree we can build from it:

![](img/11.png)

This tree predicts all of our values perfectly. It will always get a right answer on the training set, but this is equivalent to memorizing the rules of addition. While we've built our tree in such a way that it can perfectly predict the training set, the way it's constructed doesn't make sense when we take a step back.<br>

That's because the tree above is saying that:

* If you're under `22.5` years old, you have a low income
* If you're `22.5` - `37.5`, you have a high income
* If you're `37.5` - `47.5`, you have a low income
* If you're `47.5` to `55`, you have a high income
* Finally, if you're above `55`, you have a low income

These rules are very specific to the training set.<br>

Think about the problem with a real-world lens. Does it make sense to predict that someone who's `20` has a low income, someone who's `25` has a high income, and someone who's `40` has a low income? Intuitively, we know that younger people tend to earn less, middle-aged people earn more, and people who have retired earn less.<br>

Our tree has created so many age-based splits in an attempt to perfectly predict everyone's income that each split is effectively meaningless.<br>

Here's a tree that matches up with our intuition better:

![](img/12.png)

All we've done is "pruned" the tree, and removed some of the lower leaves. We've turned some of the higher-level nodes into leaves instead.<br>

The tree above makes more intuitive sense. If you're under `25`, we predict low income. If you're between `25` and `55`, we predict high income (the `.66` rounds up to 1). If you're above `55`, we predict low income.<br>

This version actually has lower accuracy on our training set, but will generalize to new examples better because it matches reality more closely.<br>

Trees overfit when they have too much depth and make overly complex rules that match the training data, but aren't able to generalize well to new data. This may seem to be a strange principle at first, but the deeper a tree is, the worse it typically performs on new data.

## Reducing Overfitting With a Shallower Tree

There are three main ways to combat overfitting:

1. "Prune" the tree after we build it to remove unnecessary leaves.
2. Use ensembling to blend the predictions of many trees.
3. Restrict the depth of the tree while we're building it.

While we'll explore all of these, we'll look at the third method first.<br>

Limiting tree depth during the building process will result in more general rules. This prevents the tree from overfitting.<br>

We can restrict tree depth by adding a few parameters when we initialize the `DecisionTreeClassifier` class:

* `max_depth` - Globally restricts how deep the tree can go
* `min_samples_split` - The minimum number of rows a node should have before it can be split; if this is set to 2, for example, then nodes with 2 rows won't be split, and will become leaves instead
* `min_samples_leaf` - The minimum number of rows a leaf must have
* `min_weight_fraction_leaf` - The fraction of input rows a leaf must have
* `max_leaf_nodes` - The maximum number of total leaves; this will cap the count of leaf nodes as the tree is being built

Some of these parameters aren't compatible, however. For example, we can't use `max_depth` and `max_leaf_nodes` together.<br>

Now that we know what to tweak, let's improve our model.

* Set `min_samples_split` to `13` when creating the `DecisionTreeClassifier`.
* Make predictions on the training set, compute the AUC, and assign it to `train_auc`.
* Make predictions on the test set, compute the AUC, and assign it to `test_auc`.

In [9]:
# Decision trees model from the last screen
clf = DecisionTreeClassifier(random_state=1,
                            min_samples_split = 13)
clf.fit(train[columns], train['high_income'])
pred = clf.predict(train[columns])

train_auc = roc_auc_score(train['high_income'], pred)
test_auc = roc_auc_score(test['high_income'],
                        clf.predict(test[columns]))

## Tweaking Parameters to Adjust AUC

By setting `min_samples_split` to `13`, we managed to boost the test AUC from `.694` to `.700`. The training set AUC decreased from `.947` to `.843`, showing that the model we built was less overfit to the training set than before:

settings|train AUC|test AUC
---|---|---
default|0.947|0.694
min_sample_split:13|0.843|0.700

Let's play around with parameters some more.

* Set `max_depth` to `7` and `min_samples_split` to `13` when creating the `DecisionTreeClassifier`.
* Make predictions on the training set, compute the AUC, and assign it to `train_auc`.
* Make predictions on the test set, compute the AUC, and assign it to `test_auc`.

In [10]:
# The first decision trees model we trained and tested
clf = DecisionTreeClassifier(random_state=1)
clf.fit(train[columns], train["high_income"])
predictions = clf.predict(test[columns])
test_auc = roc_auc_score(test["high_income"], predictions)

train_predictions = clf.predict(train[columns])
train_auc = roc_auc_score(train["high_income"], train_predictions)

print(test_auc)
print(train_auc)

0.693568175543
0.947124450144


In [12]:
# The first decision trees model we trained and tested
clf = DecisionTreeClassifier(random_state=1,
                            min_samples_split=100,
                            max_depth=2)
clf.fit(train[columns], train["high_income"])

predictions = clf.predict(test[columns])
test_auc = roc_auc_score(test["high_income"], predictions)

train_predictions = clf.predict(train[columns])
train_auc = roc_auc_score(train["high_income"], train_predictions)

print(test_auc)
print(train_auc)

0.655313848188
0.662450804216


## Underfitting In Simplistic Trees

Our accuracy went down on the last screen, relative to the screen before it:

settings|train AUC|test AUC
---|---|---
default(min_samples_split: 2, max_depth: None)|0.947|0.694
min_samples_split: 13|0.843|0.700
min_samples_split: 13, max_depth: 7|0.748|0.7744
min_samples_split: 100, max_depth: 2|0.662|0.655

This is because we're now [underfitting](http://datascience.stackexchange.com/questions/361/when-is-a-model-underfitted). Underfitting is what occurs when our model is too simple to explain the relationships between the variables.<br>

Let's go back to our tree diagram to explain underfitting.<br>

Here's the data:

```python
high_income    age    marital_status
0              20     0
0              60     2
0              40     1
1              25     1
1              35     2
1              55     1
```

And here's the "right fit" tree. This tree explains the data properly, without overfitting:

![](img/13.png)

Let's trim this tree even more to show what happens when the model isn't complex enough to explain the data:

![](img/14.png)

This model will predict that anyone under `37.5` has a high income (`.66` rounds up), and anyone over `37.5` has a low income (`.33` rounds down). It's too simple to model reality, in which people earn less while they're young, more during middle age, and less again after retirement.<br>

Therefore, this tree underfits the data and will have lower accuracy than the appropriate version.

## The Bias-Variance Tradeoff

By artificially restricting the depth of our tree, we prevent it from creating a model that's complex enough to correctly categorize some of the rows. If we don't perform the artificial restrictions, however, the tree becomes too complex, fits quirks in the data that only exist in the training set, and doesn't generalize to new data.<br>

This is known as the [bias-variance tradeoff](https://en.wikipedia.org/wiki/Bias%E2%80%93variance_tradeoff). Imagine that we take a random sample of training data and create many models. 
* If the models' predictions for the same row are far apart from each other, we have high variance. 
* Imagine this time that we take a random sample of the training data and create many models. If the models' predictions for the same row are close together but far from the actual value, then we have high bias.<br>

### High bias can cause underfitting 
-- if a model is **consistently failing to predict** the correct value, it may be that it's **too simple to model the data** faithfully.<br>

### High variance can cause overfitting. 
If a model varies its predictions significantly based on small changes in the input data, then it's likely fitting itself to quirks in the training data, **rather than making a generalizable model**.<br>

We call this the bias-variance tradeoff because decreasing one characteristic will usually increase the other. This is a limitation of all machine learning algorithms. If you'd like to read more about the tradeoff, check out [programmer Scott Fortmann-Roe's post on this topic](http://scott.fortmann-roe.com/docs/BiasVariance.html).<br>

Decision trees typically suffer from high variance. The entire structure of a decision tree can change if we make a minor alteration to its training data. **By restricting the depth of the tree, we increase the bias and decrease the variance**. If we restrict the depth too much, we increase bias to the point where it will underfit.<br>

You'll generally need to use your intuition and manually tweak parameters to get the "right" fit.

## Exploring Decision Tree Variance

We can induce variance and see what happens with a decision tree. To introduce noise into the data, we'll add a column of random values. A model with high variance (like a decision tree) will pick up on this noise, and overfit to it. This is because models with high variance are very sensitive to small changes in input data.

* Fit the classifier to the training data.

* Make predictions on the training set, compute the AUC, and assign it to `train_auc`.

* Make predictions on the test set, compute the AUC, and assign it to `test_auc`.

In [15]:
np.random.seed(1)

# Generate a column containing random numbers from 0 to 4
income["noise"] = np.random.randint(4, size=income.shape[0])

# Adjust "columns" to include the noise column
columns = ["noise", "age", "workclass", "education_num", "marital_status", "occupation", "relationship", "race", "sex", "hours_per_week", "native_country"]

# Make new train and test sets
train_max_row = math.floor(income.shape[0] * .8)
train = income.iloc[:train_max_row]
test = income.iloc[train_max_row:]

# Initialize the classifier
clf = DecisionTreeClassifier(random_state=1)
clf.fit(train[columns], train["high_income"])

predictions = clf.predict(test[columns])
test_auc = roc_auc_score(test["high_income"], predictions)

train_predictions = clf.predict(train[columns])
train_auc = roc_auc_score(train["high_income"], train_predictions)

print(test_auc)
print(train_auc)

0.691406001394
0.975076161435


## Pruning Leaves to Prevent Overfitting

As you can see above, the random noise column causes significant overfitting. Our test set accuracy decreases to `.691`, and our training set accuracy increases to `.975`.<br>

One way to prevent overfitting is to block the tree from growing beyond a certain depth (we tried this before). 

### Another technique is called pruning. 

Pruning involves 
* building a **full tree**, 
* then **removing the leaves that don't add to prediction accuracy**. 

Pruning prevents a model from becoming overly complex. It can result in a simpler model that has higher accuracy on the testing set.<br>

Data scientists use pruning less often than parameter optimization (what we just did) and ensembling. It's still an important technique, though, and we'll cover it in more depth down the line.

## Knowing When to Use Decision Trees

Let's go over the main advantages and disadvantages of using decision trees. The main advantages of using decision trees is that they're:

* Easy to interpret
* Relatively fast to fit and make predictions
* Able to handle multiple types of data
* Able to pick up nonlinearities in data, and usually fairly accurate

The main disadvantage of using decision trees is their **tendency to overfit**.<br>

Decision trees are a good choice for tasks where it's important to be able to interpret and convey why the algorithm is doing what it's doing.<br>

The most powerful way to reduce decision tree overfitting is to **create ensembles of trees**. The [random forest](https://en.wikipedia.org/wiki/Random_forest) algorithm is a popular choice for doing this. In cases where prediction accuracy is the most important consideration, random forests usually perform better.<br>

In the next mission, we'll explore the random forest algorithm in greater depth.